libbeat/common/schema/mapstriface/mapstriface_test.go

```diff
@@ -103,6 +103,122 @@ func TestConversions(t *testing.T) {
 		"test_array": []string{"a", "b", "c"},
 	}
 
-	output, _ := schema.Apply(input)
-	assert.Equal(t, output, expected)
+	result, _ := schema.Apply(input)
+	assert.Equal(t, result.Event, expected)
+}
+
+func TestOptionalField(t *testing.T) {
+	cases := []struct {
+		Description string
+		Input       map[string]interface{}
+		Schema      s.Schema
+		Expected    common.MapStr
+		ExpectError bool
+	}{
+		{
+			"missing optional field",
+			map[string]interface{}{
+				"testString": "hello",
+				"testInt":    42,
+			},
+			s.Schema{
+				"test_string": Str("testString"),
+				"test_int":    Int("testInt"),
+				"test_opt":    Bool("testOptionalInt", s.Optional),
+			},
+			common.MapStr{
+				"test_string": "hello",
+				"test_int":    int64(42),
+			},
+			false,
+		},
+		{
+			"wrong format in optional field",
+			map[string]interface{}{
+				"testInt": "hello",
+			},
+			s.Schema{
+				"test_int": Int("testInt", s.Optional),
+			},
+			common.MapStr{},
+			true,
+		},
+	}
+
+	for _, c := range cases {
+		result, err := c.Schema.Apply(c.Input)
+		if c.ExpectError {
+			assert.Error(t, err, c.Description)
+		} else {
+			assert.NoError(t, err, c.Description)
+			assert.Equal(t, c.Expected, result.Event, c.Description)
+		}
+	}
+}
+
+func TestFullFieldPathInErrors(t *testing.T) {
+	cases := []struct {
+		Description string
+		Schema      s.Schema
+		Input       map[string]interface{}
+		Expected    string
+	}{
+		{
+			"missing nested key",
+			s.Schema{
+				"a": Dict("a", s.Schema{
+					"b": Dict("b", s.Schema{
+						"c": Bool("c"),
+					}),
+				}),
+			},
+			map[string]interface{}{
+				"a": map[string]interface{}{
+					"b": map[string]interface{}{},
+				},
+			},
+			`a.b.c`,
+		},
+		{
+			"wrong nested format key",
+			s.Schema{
+				"testDict": Dict("test_dict", s.Schema{
+					"testBool": Bool("test_bool"),
+				}),
+			},
+			map[string]interface{}{
+				"test_dict": map[string]interface{}{
+					"test_bool": "foo",
+				},
+			},
+			`test_dict.test_bool`,
+		},
+		{
+			"wrong nested sub-dictionary",
+			s.Schema{
+				"testDict": Dict("test_dict", s.Schema{
+					"testDict": Dict("test_dict", s.Schema{}),
+				}),
+			},
+			map[string]interface{}{
+				"test_dict": map[string]interface{}{
+					"test_dict": "foo",
+				},
+			},
+			`test_dict.test_dict`,
+		},
+	}
+
+	for _, c := range cases {
+		_, err := c.Schema.Apply(c.Input)
+		assert.Error(t, err, c.Description)
+		assert.Contains(t, err.Error(), c.Expected, c.Description)
+
+		_, errs := c.Schema.ApplyTo(common.MapStr{}, c.Input)
+		assert.Equal(t, 1, len(errs), c.Description)
+		keyErr, ok := errs[0].(s.KeyError)
+		if assert.True(t, ok, c.Description) {
+			assert.Equal(t, c.Expected, keyErr.Key(), c.Description)
+		}
```

Comment:
`ApplyTo` can be used as a more advanced method for special cases, it still returns the full list of errors, that can be further analyzed if needed.